ProfessorAssessor/View Controllers/ProfessorViewController.m

```diff
@@ -9,7 +23,141 @@ @implementation ProfessorViewController
+    self.averageRating.value = [self.professor.averageRating doubleValue];
+}
+
+- (void)fetchCoursesAndReviews {
+    self.courses = [self sortedCourses];
+    self.selectedCourses = [NSMutableArray new];
+
+    [self displayCourseTags];
+
+    [self fetchReviews];
+}
+
+- (void)fetchReviews {
+    __weak typeof(self) weakSelf = self;
+
+    [Networker
+     fetchReviewsForProfessor:weakSelf.professor
+     forCourses:weakSelf.selectedCourses
+     withCompletion:^(NSArray<Review *> *_Nullable objects,
+                      NSError *_Nullable error) {
+        if (objects) {
+            weakSelf.reviews = objects;
+
+            [weakSelf.tableView reloadData];
+            [weakSelf.tableView.refreshControl endRefreshing];
+        }
+    }];
+}
+
+- (void)displayCourseTags { // TODO: make clear tags button
+    TTGTextTagCollectionView *courseSelectionView = [self setUpTagCollectionView];
+    [self.tableView addSubview:courseSelectionView];
+
+    TTGTextTagStyle *unselectedStyle = [self
+                                        setUpTagStyleWithColor:[UIColor lightGrayColor]];
+    TTGTextTagStyle *selectedStyle = [self
+                                      setUpTagStyleWithColor:[UIColor systemTealColor]];
+
+    for (Course *course in self.courses) {
+        TTGTextTag *courseTag = [TTGTextTag
+                                 tagWithContent:[TTGTextTagStringContent
+                                                 contentWithText:course.name]
+                                 style:unselectedStyle];
+
+        [courseTag setSelectedStyle:selectedStyle];
+        [courseTag setSelected:YES];
+
+        [self.selectedCourses addObject:course];
+
+        [courseSelectionView addTag:courseTag];
+    }
+}
+
+- (TTGTextTagCollectionView *)setUpTagCollectionView {
+    TTGTextTagCollectionView *tagCollectionView =
+    [[TTGTextTagCollectionView alloc]
+     initWithFrame:CGRectMake(20, 127, 350, 64)];
+
+    tagCollectionView.delegate = self;
+    tagCollectionView.alignment = TTGTagCollectionAlignmentCenter;
+
+    return tagCollectionView;
+}
+
+- (TTGTextTagStyle *)setUpTagStyleWithColor:(UIColor *)color {
+    TTGTextTagStyle *style = [TTGTextTagStyle new];
+
+    style.backgroundColor = color;
+    style.shadowColor = [UIColor clearColor];
+    style.shadowOffset = CGSizeMake(0, 0);
+    style.shadowRadius = 0;
+    style.borderWidth = 0;
+    style.extraSpace = CGSizeMake(10, 5);
+
+    return style;
+}
+
+- (void)textTagCollectionView:(TTGTextTagCollectionView *)textTagCollectionView
+                    didTapTag:(TTGTextTag *)tag
+                      atIndex:(NSUInteger)index { // TODO: consider not fetching new reviews unless user refreshes (selecting/unselecting tag update tableView only, not reviews)
+    Course *course = self.courses[index];
+
+    if ([self.selectedCourses containsObject:course]) {
+        [self.selectedCourses removeObject:course];
+
+        [self fetchReviews];
```

Comment:
This seems like there would be a known number of reviews for a professor. Does some kind of pre-fetching / caching make sense here?

Reply:
This seems like a LOT of fetches.